eoscompanion/routes.py

```diff
@@ -171,6 +174,102 @@ def companion_app_server_feed_route(server, msg, path, query, *args):
+
+    The querystring param "path" indicates the URI encoded path to the
+    internal resource (it may be a GResource or a file).
+    '''
+    del args
+
+    resource_uri = Soup.URI.decode(query['uri'])
+    resource_file = Gio.File.new_for_uri(resource_uri)
+
+    resource_suffix = os.path.splitext(resource_uri)[1]
+    return_content_type = _SUFFIX_CONTENT_TYPES.get(resource_suffix, None)
+
+    if return_content_type is None:
+        json_response(msg, {
+            'status': 'error',
+            'error': {
+                'domain': GLib.quark_to_string(EosCompanionAppService.error_quark()),
+                'code': EosCompanionAppService.Error.FAILED,
+                'detail': {
+                    'server_error': (
+                        'Don\'t know content type for suffix, {}'.format(resource_suffix)
+                    )
+                }
+            }
+        })
+        return
+
+    def _on_read_input_stream(_, result):
+        '''Callback for when we finish reading the input stream.'''
+        try:
+            content_bytes = EosCompanionAppService.finish_load_all_in_stream_to_bytes(result)
+        except GLib.Error as error:
+            json_response(msg, {
+                'status': 'error',
+                'error': {
+                    'domain': GLib.quark_to_string(EosCompanionAppService.error_quark()),
+                    'code': EosCompanionAppService.Error.FAILED,
+                    'detail': {
+                        'server_error': str(error)
+                    }
+                }
+            })
+            server.unpause_message(msg)
+            return
+
+        custom_response(msg, return_content_type, content_bytes)
+        server.unpause_message(msg)
+
+    def _on_read_resource(src, result):
+        '''Callback for once we finish reading the resource.'''
+        try:
+            input_stream = src.read_finish(result)
+        except GLib.Error as error:
+            if error.matches(Gio.io_error_quark(), Gio.IOErrorEnum.NOT_FOUND):
+                not_found_response(msg, path)
+            else:
+                json_response(msg, {
+                    'status': 'error',
+                    'error': {
+                        'domain': GLib.quark_to_string(EosCompanionAppService.error_quark()),
+                        'code': EosCompanionAppService.Error.FAILED,
+                        'detail': {
+                            'server_error': str(error)
+                        }
+                    }
+                })
+
+            server.unpause_message(msg)
+            return
+
+        EosCompanionAppService.load_all_in_stream_to_bytes(input_stream,
+                                                           chunk_size=BYTE_CHUNK_SIZE,
+                                                           cancellable=None,
+                                                           callback=_on_read_input_stream)
+
+    resource_file.read_async(io_priority=GLib.PRIORITY_DEFAULT,
```

Comment:
Is this a resource file in the sense of a GResource? If so, we could probably just use the `GResource` helpers to get the bytes directly instead of doing this.

Reply:
Its a "resource file" in the sense that it can be an arbitrary URI accessible to GIO (so file:/// included).

I think later on I'd like to see if we can share some of the logic that we also use in /content_data for this, but not right now.